core/src/main/scala/kafka/controller/ReplicaStateMachine.scala

```diff
@@ -122,178 +139,214 @@ class ReplicaStateMachine(controller: KafkaController, stateChangeLogger: StateC
+        val updatedLeaderIsrAndControllerEpochs = removeReplicasFromIsr(replicaId, replicasToRemoveFromIsr.map(_.topicAndPartition))
+        updatedLeaderIsrAndControllerEpochs.foreach { case (partition, leaderIsrAndControllerEpoch) =>
+          if (!topicDeletionManager.isPartitionToBeDeleted(partition)) {
+            val recipients = controllerContext.partitionReplicaAssignment(partition).filterNot(_ == replicaId)
+            controllerBrokerRequestBatch.addLeaderAndIsrRequestForBrokers(recipients,
+              partition.topic,
+              partition.partition,
+              leaderIsrAndControllerEpoch,
+              controllerContext.partitionReplicaAssignment(partition), isNew = false)
+          }
+          replicaState.put(PartitionAndReplica(partition.topic, partition.partition, replicaId), OfflineReplica)
+        }
+      case ReplicaDeletionStarted =>
+        validReplicas.foreach { replica =>
+          replicaState.put(replica, ReplicaDeletionStarted)
+          controllerBrokerRequestBatch.addStopReplicaRequestForBrokers(Seq(replicaId),
+            replica.topic,
+            replica.partition,
+            deletePartition = true,
+            callbacks.stopReplicaResponseCallback)
+        }
+      case ReplicaDeletionIneligible =>
+        validReplicas.foreach(replica => replicaState.put(replica, ReplicaDeletionIneligible))
+      case ReplicaDeletionSuccessful =>
+        validReplicas.foreach(replica => replicaState.put(replica, ReplicaDeletionSuccessful))
+      case NonExistentReplica =>
+        validReplicas.foreach { replica =>
+          val currentAssignedReplicas = controllerContext.partitionReplicaAssignment(replica.topicAndPartition)
+          controllerContext.partitionReplicaAssignment.put(replica.topicAndPartition, currentAssignedReplicas.filterNot(_ == replica.replica))
+          replicaState.remove(replica)
+        }
+    }
+  }
+  private def removeReplicasFromIsr(replicaId: Int, partitions: Seq[TopicAndPartition]) = {
+    var results = Map.empty[TopicAndPartition, LeaderIsrAndControllerEpoch]
+    var remaining = partitions
+    while (remaining.nonEmpty) {
+      val (successfulRemovals, removalsToRetry, failedRemovals) = doRemoveReplicasFromIsr(replicaId, remaining)
+      results ++= successfulRemovals
+      remaining = removalsToRetry
+      failedRemovals.foreach { case (partition, e) =>
+        val replica = PartitionAndReplica(partition.topic, partition.partition, replicaId)
+        logFailedStateChange(replica, replicaState(replica), OfflineReplica, e)
       }
     }
-    catch {
-      case t: Throwable =>
-        stateChangeLog.error(s"Initiated state change of replica $replicaId for partition $topicAndPartition from " +
-          s"$currState to $targetState failed", t)
+    results
+  }
+
+  private def doRemoveReplicasFromIsr(replicaId: Int, partitions: Seq[TopicAndPartition]) = {
```

Comment:
Could we explicitly define the return type explicitly of this method and describe a bit the return type and the method? Ditto for the method below.